### PR TITLE
ForTA#20

### PR DESCRIPTION
**Pull request by VladykaIhor**

### CODE REVIEW

File: onlineshoptomcat/onlineshoptomcat.iml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Sofasmile) [inline onlineshoptomcat/onlineshoptomcat.iml]: add .iml files and some packages to .gitignore

File: onlineshoptomcat/pom.xml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>3.1.0</version>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+</project>
```
Comment (Sofasmile) [inline onlineshoptomcat/pom.xml]: add a blank line

File: onlineshoptomcat/src/main/java/model/Product.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    private Long id = 0L;
+    private String name;
+    private String decription;
+    private Double price;
+
+    public Product(String name, String decription, Double price) {
+        this.name = name;
+        this.decription = decription;
+        this.price = price;
+        id++;
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/model/Product.java]: does it work correctly?

File: onlineshoptomcat/src/main/java/dao/UserDao.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package dao;
+
+import model.User;
+
+import java.util.List;
+
+public interface UserDao {
+
+    void addUser(User user);
+    List<User> getAll();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/dao/UserDao.java]: cs

File: onlineshoptomcat/src/main/java/dao/impl/ProductDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package dao.impl;
+
+import dao.ProductDao;
+import model.Product;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ProductDaoImpl implements ProductDao {
+    private static List<Product> productList = new ArrayList<Product>();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/dao/impl/ProductDaoImpl.java]:
```suggestion
    private static List<Product> products = new ArrayList<Product>();
```

File: onlineshoptomcat/src/main/java/dao/impl/UserDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import dao.UserDao;
+import model.User;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserDaoImpl implements UserDao {
+
+    private static List<User> userList = new ArrayList<User>();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/dao/impl/UserDaoImpl.java]:
```suggestion
    private static List<User> users = new ArrayList<User>();
```

File: onlineshoptomcat/src/main/java/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package service;
+
+import model.User;
+
+import java.util.List;
+
+public interface UserService {
+
+    void addUser(User user);
+    List<User> getAll();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/service/UserService.java]: cs

File: onlineshoptomcat/src/main/java/service/impl/ProductServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import dao.ProductDao;
+import dao.impl.ProductDaoImpl;
+import model.Product;
+import service.ProductService;
+
+import java.util.List;
+
+public class ProductServiceImpl implements ProductService {
+
+    private static final ProductDao productDao = new ProductDaoImpl();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/service/impl/ProductServiceImpl.java]: we need to use factory here

File: onlineshoptomcat/src/main/java/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+import dao.UserDao;
+import dao.impl.UserDaoImpl;
+import model.User;
+import service.UserService;
+
+import java.util.List;
+
+public class UserServiceImpl implements UserService {
+
+    private static final UserDao userDao = new UserDaoImpl();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/service/impl/UserServiceImpl.java]: we need to use factory here

File: onlineshoptomcat/web/WEB-INF/web.xml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<web-app xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_4_0.xsd"
+         version="4.0">
+
+</web-app>
```
Comment (Sofasmile) [inline onlineshoptomcat/web/WEB-INF/web.xml]: add a blank line

File: onlineshoptomcat/src/main/java/controller/AllUserServlet.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.List;
+
+@WebServlet("/users")
+public class AllUserServlet extends HttpServlet {
+
+    private static final UserService userService = new UserServiceImpl();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/controller/AllUserServlet.java]: we need to use factory here

File: onlineshoptomcat/src/main/java/controller/ProductAddServlet.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String description = req.getParameter("description");
+        Double price = Double.parseDouble(req.getParameter("price"));
+
+        Product product = new Product(name, description, price);
+        productService.addProduct(product);
+        doGet(req, resp);
+    }
+}
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/controller/ProductAddServlet.java]: blank line

File: onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@WebServlet(value = "/register")
+public class UserRegistrationServlet extends HttpServlet {
+
+    private static final UserService userService = UserServiceFactory.getUserService();
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java]: https://mate-academy.github.io/style-guides/java/java.html#s4.4-column-limit

File: onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        req.getRequestDispatcher("register.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String email = req.getParameter("email");
+        String password = req.getParameter("password");
+        String repeatPassword = req.getParameter("repeatPassword");
+        if (password.equals(repeatPassword)) {
+            User user = new User(1L, email, password);
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java]: 1L ?

File: onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+            userService.addUser(user);
+            resp.sendRedirect("/");
+        } else {
+            req.setAttribute("error", "The repeated passwords aren't equal");
+            req.getRequestDispatcher("register.jsp").forward(req, resp);
+        }
+
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java]: cs

File: onlineshoptomcat/web/product.jsp [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+<br>
+<center>
+    <form action="product" method="post">
+        Name <input name="name" type="name"/> <br>
+        Description <input name="description" type="description"> <br>
+        Price <input name="price" type="price"> <br>
+        <button type="submit">Add Product</button>
+    </form>
+</center>
+
```
Comment (Sofasmile) [inline onlineshoptomcat/web/product.jsp]: cs

File: onlineshoptomcat/web/users.jsp [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        printWriter.write("<tr>");
+        printWriter.write("<td>" + user.getEmail() + "</td>");
+        printWriter.write("<td>" + user.getPassword() + "</td>");
+        printWriter.write("</tr>");
+    }
+    printWriter.write("</center>");
+%>
+
+</table>
+
```
Comment (Sofasmile) [inline onlineshoptomcat/web/users.jsp]: cs

File: onlineshoptomcat/out/artifacts/onlineshoptomcat_war_exploded/META-INF/MANIFEST.MF [hunk 18]
```diff
@@ -0,0 +1,5 @@
+Manifest-Version: 1.0
```
Comment (Sofasmile) [inline onlineshoptomcat/out/artifacts/onlineshoptomcat_war_exploded/META-INF/MANIFEST.MF]: this file better add to .gitignore too

File: onlineshoptomcat/src/main/java/controller/ProductAddServlet.java (hunk 12 above)
Comment (Sofasmile) [reply onlineshoptomcat/src/main/java/controller/ProductAddServlet.java]: ?

File: onlineshoptomcat/out/artifacts/onlineshoptomcat_war_exploded/META-INF/MANIFEST.MF (hunk 18 above)
Comment (VladykaIhor) [reply onlineshoptomcat/out/artifacts/onlineshoptomcat_war_exploded/META-INF/MANIFEST.MF]: done

File: onlineshoptomcat/src/main/java/controller/AllUserServlet.java (hunk 11 above)
Comment (Sofasmile) [reply onlineshoptomcat/src/main/java/controller/AllUserServlet.java]: What about factory?

File: onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java (hunk 14 above)
Comment (Sofasmile) [reply onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java]: Why we set id as 1L all time?

File: onlineshoptomcat/src/main/java/model/Product.java (hunk 3 above)
Comment (Sofasmile) [reply onlineshoptomcat/src/main/java/model/Product.java]: what about increment id? 
test which id we will have in our storage after adding 3-5 elements?

File: onlineshoptomcat/src/main/java/service/impl/ProductServiceImpl.java (hunk 8 above)
Comment (Sofasmile) [reply onlineshoptomcat/src/main/java/service/impl/ProductServiceImpl.java]: create factory

File: onlineshoptomcat/src/main/java/controller/AllUserServlet.java (hunk 11 above)
Comment (VladykaIhor) [reply onlineshoptomcat/src/main/java/controller/AllUserServlet.java]: This servlet was created by German. Fixed it anyways

File: onlineshoptomcat/src/main/java/service/impl/UserServiceImpl.java (hunk 9 above)
Comment (VladykaIhor) [reply onlineshoptomcat/src/main/java/service/impl/UserServiceImpl.java]: Similar thing - A UserServiceImpl class was created by German as well, so I've created it literally by a template

File: onlineshoptomcat/src/main/java/model/Product.java (hunk 3 above)
Comment (VladykaIhor) [reply onlineshoptomcat/src/main/java/model/Product.java]: fixed

File: onlineshoptomcat/src/main/java/service/UserService.java (hunk 7 above)
Comment (VladykaIhor) [reply onlineshoptomcat/src/main/java/service/UserService.java]: fixed

File: onlineshoptomcat/src/main/java/dao/UserDao.java (hunk 4 above)
Comment (VladykaIhor) [reply onlineshoptomcat/src/main/java/dao/UserDao.java]: fixed

File: onlineshoptomcat/src/main/java/controller/AllUserServlet.java [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.List;
+
+@WebServlet("/users")
+public class AllUserServlet extends HttpServlet {
+
+    private static final UserService userService = UserServiceFactory.getUserService();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/controller/AllUserServlet.java]: cs - naming in all servlets

File: onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+        req.getRequestDispatcher("register.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+        throws ServletException, IOException {
+        String email = req.getParameter("email");
+        String password = req.getParameter("password");
+        String repeatPassword = req.getParameter("repeatPassword");
+        int id = userService.getAll().size();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/controller/UserRegistrationServlet.java]: what does this line do?

File: onlineshoptomcat/src/main/java/dao/ProductDao.java [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package dao;
+
+import model.Product;
+
+import java.util.List;
+
+public interface ProductDao {
+
+    void addProduct(Product product);
+
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/dao/ProductDao.java]: better add line between methods in ALL interfaces or remove it in ALL.

File: onlineshoptomcat/src/main/java/factory/ProductDaoFactory.java [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+import dao.ProductDao;
+import dao.impl.ProductDaoImpl;
+
+public class ProductDaoFactory {
+    private static ProductDao productDaoInstance;
+
+    public ProductDaoFactory() {
+    }
+
+    public static ProductDao getProductDaoInstance() {
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/factory/ProductDaoFactory.java]: better rename to getInstance() in all factories

File: onlineshoptomcat/src/main/java/factory/ProductServiceFactory.java [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+import service.ProductService;
+import service.UserService;
+import service.impl.ProductServiceImpl;
+import service.impl.UserServiceImpl;
+
+public class ProductServiceFactory {
+    public ProductServiceFactory() {
+    }
+
+    private static ProductService instance;
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/factory/ProductServiceFactory.java]: field in this class has name 'instance', but in ProductDaoFactory - 'productDaoInstance'
try to use single name in all factories. will be good if name is 'instance'

File: onlineshoptomcat/src/main/java/service/impl/ProductServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+import dao.ProductDao;
+import factory.ProductDaoFactory;
+import model.Product;
+import service.ProductService;
+
+import java.util.List;
+
+public class ProductServiceImpl implements ProductService {
+
+    private static final ProductDao productDao = ProductDaoFactory.getProductDaoInstance();
```
Comment (Sofasmile) [inline onlineshoptomcat/src/main/java/service/impl/ProductServiceImpl.java]: cs - naming

**Pull request closed without merging**